examples/speculative/speculative.cpp

```diff
@@ -196,66 +202,169 @@ int main(int argc, char ** argv) {
         int i_dft  = 0;
         int s_keep = 0;
 
-        while (true) {
-            LOG("sampling target: s_keep = %3d, i_dft = %3d, i_batch_tgt = %3d\n", s_keep, i_dft, drafts[s_keep].i_batch_tgt[i_dft]);
+        llama_token token_id;
+        std::string token_str;
 
-            // sample from the target model
-            llama_token id = llama_sampling_sample(ctx_sampling, ctx_tgt, NULL, drafts[s_keep].i_batch_tgt[i_dft]);
+        // loop until we fail to accept a drafted token or we run out of drafted tokens
+        while (true) {
 
-            llama_sampling_accept(ctx_sampling, ctx_tgt, id, true);
+            // check if the target token matches any of the drafts
+            // for stochastic sampling, attempt to match the token with the drafted tokens
+            {
+                bool accept = false;
+                if (params.sparams.temp > 0) {
+                    // stochastic verification
 
-            //LOG("last: %s\n", LOG_TOKENS_TOSTR_PRETTY(ctx_tgt, ctx_sampling->prev).c_str());
+                    llama_token_data_array dist_tgt = llama_sampling_probability_distribution(ctx_sampling, ctx_tgt, NULL, drafts[s_keep].i_batch_tgt[i_dft]);
+                    float p_tgt, p_dft;
 
-            const std::string token_str = llama_token_to_piece(ctx_tgt, id);
+                    // GGML_ASSERT(dist_tgt.size() == dist_dft.size());
+                    for (int s = 0; s < n_seq_dft; ++s) {
+                        if (!drafts[s].active) {
+                            continue;
+                        }
+                        if (i_dft >= (int) drafts[s].tokens.size()) {
+                            drafts[s].active = false;
+                            continue;
+                        }
+                        if (accept) {
+                            // if we already accepted a token, we can skip the rest
+                            if (drafts[s].tokens[i_dft] != drafts[s_keep].tokens[i_dft]) {
+                                drafts[s].active = false;
+                            }
+                            continue;
+                        }
 
-            if (!params.use_color) {
-                printf("%s", token_str.c_str());
-            }
+                        float r = rand() / (float) RAND_MAX;
+                        llama_token_data_array dist_dft = drafts[s].dist[i_dft];
+                        // acquire the probability of the token from the draft model
+                        for (int i = 0; i < dist_tgt.size; i++) {
+                            if (dist_tgt.data[i].id == drafts[s].tokens[i_dft]) {
+                                p_tgt = dist_tgt.data[i].p;
+                            }
+                            if (dist_dft.data[i].id == drafts[s].tokens[i_dft]) {
+                                p_dft = dist_dft.data[i].p;
+                            }
+                            if (p_tgt && p_dft) {
```

Comment:
Reading p_tgt without initialization will definitely produce undefined behavior, and it will almost never contain zero (it's garbage data from the stack).